Getting and Cleaning Data
====

### 01 - Introduction

Lending Club is a peer to peer lending company based in the United States, in which investors provide funds for potential borrowers and investors earn a profit depending on the risk they take (the borrowers credit score). Lending Club provides the "bridge" between investors and borrowers.

For more information about the company, please check out the official [website](www.lendingclub.com)

__About the data__

Lending club provides several [csv files](https://www.lendingclub.com/info/download-data.action) that contain complete loan data for all loans issued from 2007 to last 2017 quarter, including the current loan status (Current, Late, Fully Paid, etc.) and latest payment information. Some additional features such as credit scores, number of finance inquiries, zip codes, states, or collections have been included among others.

Lending club provides a Data Dictionary that includes definitions for all the data attributes included in the Historical data file and the In Funding data file.

In this notebook we are going to get familiar with the data following the below plan of action:
* Explore files using the shell.
* Concatenate all files in just one dataset.
* Select columns of interest using the Lending Club Data Dictionary in XLM format.

Let's get started!!!

### 02 - Explore Files

__Libraries__

In [1]:
#remove warnigns
import warnings
warnings.filterwarnings("ignore")

import os

import numpy as np
import pandas as pd

We are going to use the shell for a first data exploration and getting familiar with the data:

In [2]:
raw_data_path = "/media/juanan/DATA/loan_data_analysis/data/raw"

In [3]:
!ls -lh {raw_data_path}

total 1,4G
-rw-r--r-- 1 juanan juanan 104M may  2 12:24 LoanStats_2016Q1.csv
-rw-r--r-- 1 juanan juanan  76M may  2 12:42 LoanStats_2016Q2.csv
-rw-r--r-- 1 juanan juanan  77M may  2 12:56 LoanStats_2016Q3.csv
-rw-r--r-- 1 juanan juanan  81M may  2 13:13 LoanStats_2016Q4.csv
-rw-r--r-- 1 juanan juanan  76M may  2 13:29 LoanStats_2017Q1.csv
-rw-r--r-- 1 juanan juanan  83M may  2 13:47 LoanStats_2017Q2.csv
-rw-r--r-- 1 juanan juanan  96M may  2 14:07 LoanStats_2017Q3.csv
-rw-r--r-- 1 juanan juanan  93M may  2 14:27 LoanStats_2017Q4.csv
-rw-r--r-- 1 juanan juanan  41M may  2 14:34 LoanStats3a.csv
-rw-r--r-- 1 juanan juanan 155M may  2 15:04 LoanStats3b.csv
-rw-r--r-- 1 juanan juanan 180M may  2 15:38 LoanStats3c.csv
-rw-r--r-- 1 juanan juanan 319M may  2 16:37 LoanStats3d.csv


There are 12 files with a total disk size of 1.4 GB. All of them have been downloaded from [lending-club](https://www.lendingclub.com/info/download-data.action) web site. `LoanStats3a` file corresponds to the loans issued from 2007 to 2011, `LoanStats3b` from 2012 to 2013 and `LoanStats3c` and `LoanStats3d` to the years 2014 and 2015 respectively. From 2016, the loans have been registered by quarter (files named `LoanStats_YEARQUARTER.csv`)

Let's take a look at the first five observations to one of the csv files, for example, the loans issued in the 3rd quarter of the last year:

In [4]:
!head -n 4 {raw_data_path}/LoanStats_2017Q3.csv

"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","open_il_12m","open_il_24m","mths_since_rcnt_il","total_bal_il","il_util","open_rv_

Let's count the number of lines for each file using the shell:

In [5]:
list_dir = !ls {raw_data_path}

for file in list_dir:
    observations = !cat {raw_data_path}/{file} | wc -l
    print("{} has {} lines".format(file, observations[0]))

LoanStats_2016Q1.csv has 133892 lines
LoanStats_2016Q2.csv has 97859 lines
LoanStats_2016Q3.csv has 99125 lines
LoanStats_2016Q4.csv has 103551 lines
LoanStats_2017Q1.csv has 96784 lines
LoanStats_2017Q2.csv has 105456 lines
LoanStats_2017Q3.csv has 122706 lines
LoanStats_2017Q4.csv has 118653 lines
LoanStats3a.csv has 42543 lines
LoanStats3b.csv has 188186 lines
LoanStats3c.csv has 235634 lines
LoanStats3d.csv has 421100 lines


### 03 - Concatenate all files

We are going to concatenate all the previous files in just one dataframe:

In [6]:
%%time
list_ = []
for file in list_dir:
    print("Processing {} file".format(file))
    full_path = os.path.join(raw_data_path, file)
    df = pd.read_csv(full_path, sep = ",")
    list_.append(df)
    
loans = pd.concat(list_)

Processing LoanStats_2016Q1.csv file
Processing LoanStats_2016Q2.csv file
Processing LoanStats_2016Q3.csv file
Processing LoanStats_2016Q4.csv file
Processing LoanStats_2017Q1.csv file
Processing LoanStats_2017Q2.csv file
Processing LoanStats_2017Q3.csv file
Processing LoanStats_2017Q4.csv file
Processing LoanStats3a.csv file
Processing LoanStats3b.csv file
Processing LoanStats3c.csv file
Processing LoanStats3d.csv file
CPU times: user 25.6 s, sys: 1.91 s, total: 27.5 s
Wall time: 27.8 s


Let's take a look at the first 10 rows and check up the shape of the dataframe in the raw format:

In [8]:
loans = pd.read_csv("/media/juanan/DATA/loan_data_analysis/data/clean/accepted_2007_to_2017Q3.csv.gz")

In [9]:
loans.head(10)

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  38098114        NaN    15000.0      15000.0          15000.0   60 months   
1  36805548        NaN    10400.0      10400.0          10400.0   36 months   
2  37842129        NaN    21425.0      21425.0          21425.0   60 months   
3  37612354        NaN    12800.0      12800.0          12800.0   60 months   
4  37662224        NaN     7650.0       7650.0           7650.0   36 months   
5  37822187        NaN     9600.0       9600.0           9600.0   36 months   
6  37741884        NaN     2500.0       2500.0           2500.0   36 months   
7  37854444        NaN    16000.0      16000.0          16000.0   60 months   
8  36804663        NaN    23325.0      23325.0          23325.0   36 months   
9  37642222        NaN     5250.0       5250.0           5250.0   36 months   

   int_rate  installment grade sub_grade       ...        \
0     12.39       336.64     C        C1       ...         
1      6.99       321.08     A        A3       ...         
2     15.59       516.36     D        D1       ...         
3     17.14       319.08     D        D4       ...         
4     13.66       260.20     C        C3       ...         
5     13.66       326.53     C        C3       ...         
6     11.99        83.03     B        B5       ...         
7     11.44       351.40     B        B4       ...         
8     14.31       800.71     C        C4       ...         
9     11.44       172.98     B        B4       ...         

  hardship_payoff_balance_amount hardship_last_payment_amount  \
0                            NaN                          NaN   
1                            NaN                          NaN   
2                            NaN                          NaN   
3                            NaN                          NaN   
4                            NaN                          NaN   
5                            NaN                          NaN   
6                            NaN                          NaN   
7                            NaN                          NaN   
8                            NaN                          NaN   
9                            NaN                          NaN   

  disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
0                Cash                     N                       NaN   
1                Cash                     N                       NaN   
2                Cash                     N                       NaN   
3                Cash                     N                       NaN   
4                Cash                     N                       NaN   
5                Cash                     N                       NaN   
6                Cash                     N                       NaN   
7                Cash                     N                       NaN   
8                Cash                     N                       NaN   
9                Cash                     N                       NaN   

  settlement_status settlement_date settlement_amount settlement_percentage  \
0               NaN             NaN               NaN                   NaN   
1               NaN             NaN               NaN                   NaN   
2               NaN             NaN               NaN                   NaN   
3               NaN             NaN               NaN                   NaN   
4               NaN             NaN               NaN                   NaN   
5               NaN             NaN               NaN                   NaN   
6               NaN             NaN               NaN                   NaN   
7               NaN             NaN               NaN                   NaN   
8               NaN             NaN               NaN                   NaN   
9               NaN             NaN               NaN                   NaN   

  settlement_term  
0             NaN  
1             NaN  
2             NaN  
3       

In [10]:
loans.shape

(1646801, 150)

We have 150 variables and 1646801 observations in the dataframe. But, as we can see looking at the head of the data, it seems that there are a lot of NaN columns we have to deal with. We will do it later, in the pre-processing stage.

### 04 - Select columns of interest

In this section, we are going to use the data dictionary that lending club provides at their webpage. This way, we will get only the description of the features in our current loans dataset. 

To decide whether to keep a feature or not is one of the most critical parts of the project since we have some information in the dataset related to time or previous payments that we will not have for new borrowers and they reveal much information about loan status. For example, if `debt_settlement_flag` is "Y", it implies that the borrower charged off, or if `total_pymnt` is greater than `loan_amnt`, then the loan must be paid. For this reason, only variables that don't update over time will be kept. __We can't use this kind of information for new lending club users.__

In [11]:
columns_of_interest = ['num_bc_sats', 'num_rev_tl_bal_gt_0', 'grade', 'avg_cur_bal', 'pub_rec_bankruptcies', 
                       'num_rev_accts', 'tax_liens', 'funded_amnt_inv', 'delinq_2yrs', 'total_bal_ex_mort',
                       'pct_tl_nvr_dlq', 'disbursement_method', 'fico_range_low', 'verification_status', 'delinq_amnt',
                       'purpose', 'emp_title', 'zip_code', 'loan_amnt', 'installment', 
                       'fico_range_high', 'annual_inc', 'term', 'int_rate', 'emp_length',
                       'revol_bal', 'application_type', 'num_bc_tl', 'num_sats', 'tot_hi_cred_lim', 
                       'tot_coll_amt', 'initial_list_status', 'bc_open_to_buy', 'total_bc_limit', 
                       'open_acc', 'revol_util', 'pub_rec', 'funded_amnt', 'num_il_tl', 
                       'addr_state', 'num_accts_ever_120_pd', 'total_il_high_credit_limit', 'bc_util', 'percent_bc_gt_75', 
                       'sub_grade', 'mort_acc', 'num_op_rev_tl', 'dti', 'home_ownership', 'loan_status']

In [32]:
loans = loans[sorted(columns_of_interest)]

Let's use the Data Dictionary:

In [15]:
lc_dict = pd.read_excel("/media/juanan/DATA/loan_data_analysis/data/LCDataDictionary.xlsx")

In [16]:
lc_dict.head()

LoanStatNew                                        Description
0        acc_now_delinq  The number of accounts on which the borrower i...
1  acc_open_past_24mths         Number of trades opened in past 24 months.
2            addr_state  The state provided by the borrower in the loan...
3              all_util              Balance to credit limit on all trades
4            annual_inc  The self-reported annual income provided by th...

In [17]:
features_intersection = pd.DataFrame(list(set(lc_dict['LoanStatNew']).intersection(set(loans.columns))))
features_intersection.columns = ["LoanStatNew"]

In [18]:
features_dict = pd.merge(features_intersection, lc_dict, on=['LoanStatNew'])

The meanings of the variables of interest are shown below:

In [36]:
features_dict = features_dict.sort_values("LoanStatNew").reset_index().drop("index", axis = 1)
features_dict

LoanStatNew  \
0   addr_state                   
1   annual_inc                   
2   application_type             
3   avg_cur_bal                  
4   bc_open_to_buy               
5   bc_util                      
6   delinq_2yrs                  
7   delinq_amnt                  
8   disbursement_method          
9   dti                          
10  emp_length                   
11  emp_title                    
12  fico_range_high              
13  fico_range_low               
14  funded_amnt                  
15  funded_amnt_inv              
16  grade                        
17  home_ownership               
18  initial_list_status          
19  installment                  
20  int_rate                     
21  loan_amnt                    
22  loan_status                  
23  mort_acc                     
24  num_accts_ever_120_pd        
25  num_bc_sats                  
26  num_bc_tl                    
27  num_il_tl                    
28  num_op_rev_tl                
29  num_rev_accts                
30  num_rev_tl_bal_gt_0          
31  num_sats                     
32  open_acc                     
33  pct_tl_nvr_dlq               
34  percent_bc_gt_75             
35  pub_rec                      
36  pub_rec_bankruptcies         
37  purpose                      
38  revol_bal                    
39  revol_util                   
40  sub_grade                    
41  tax_liens                    
42  term                         
43  tot_coll_amt                 
44  tot_hi_cred_lim              
45  total_bal_ex_mort            
46  total_bc_limit               
47  total_il_high_credit_limit   
48  verification_status          
49  zip_code                     

                                                                                                                                                                                                 Description  
0   The state provided by the borrower in the loan application                                                                                                                                                
1   The self-reported annual income provided by the borrower during registration.                                                                                                                             
2   Indicates whether the loan is an individual application or a joint application with two co-borrowers                                                                                                      
3   Average current balance of all accounts                                                                                                                                                                   
4   Total open to buy on revolving bankcards.                                                                                                                                                                 
5   Ratio of total current balance to high credit/credit limit for all bankcard accounts.                                                                                                                     
6   The number of 30+ days past-due incidences of delinquency in the borrower's credit file for the past 2 years                                                                                              
7   The past-due amount owed for the accounts on which the borrower is now delinquent.                                                                                                                        
8   The method by which the borrower receives their loan. Possible values are: CASH, DIRECT_PAY                                                                                                               
9   A ratio calculated using the borrower’s total monthly debt payments on the total debt obligations, excluding mortgage and the requested LC loan, divided by the borrower’s self-reported monthly income.  
10  Emp

The new dataset, a little bit smaller:

In [37]:
loans.head()

addr_state  annual_inc application_type  avg_cur_bal  bc_open_to_buy  \
0  VA         78000.0     Individual       29828.0      9525.0           
1  CA         58000.0     Individual       9536.0       7599.0           
2  MO         63800.0     Individual       4232.0       324.0            
3  CA         125000.0    Individual       32727.0      0.0              
4  AZ         50000.0     Individual       5857.0       332.0            

   bc_util  delinq_2yrs  delinq_amnt disbursement_method    dti emp_length  \
0  4.7      0.0          0.0          Cash                12.03  10+ years   
1  41.5     0.0          0.0          Cash                14.92  8 years     
2  97.8     0.0          0.0          Cash                18.49  6 years     
3  103.2    1.0          0.0          Cash                8.31   10+ years   
4  93.2     0.0          0.0          Cash                34.81  < 1 year    

                         emp_title  fico_range_high  fico_range_low  \
0  MANAGEMENT                       754.0            750.0            
1  Truck Driver Delivery Personel   714.0            710.0            
2  Programming Analysis Supervisor  689.0            685.0            
3  Senior Sales Professional        669.0            665.0            
4  Technical Specialist             689.0            685.0            

   funded_amnt  funded_amnt_inv grade home_ownership initial_list_status  \
0  15000.0      15000.0          C     RENT           w                    
1  10400.0      10400.0          A     MORTGAGE       w                    
2  21425.0      21425.0          D     RENT           w                    
3  12800.0      12800.0          D     MORTGAGE       w                    
4  7650.0       7650.0           C     RENT           f                    

   installment  int_rate  loan_amnt  loan_status  mort_acc  \
0  336.64       12.39     15000.0    Fully Paid   0.0        
1  321.08       6.99      10400.0    Charged Off  1.0        
2  516.36       15.59     21425.0    Fully Paid   0.0        
3  319.08       17.14     12800.0    Current      5.0        
4  260.20       13.66     7650.0     Charged Off  0.0        

   num_accts_ever_120_pd  num_bc_sats  num_bc_tl  num_il_tl  num_op_rev_tl  \
0  0.0                    1.0          2.0        8.0        5.0             
1  4.0                    7.0          18.0       2.0        14.0            
2  1.0                    3.0          12.0       16.0       5.0             
3  1.0                    3.0          5.0        1.0        5.0             
4  0.0                    1.0          4.0        12.0       4.0             

   num_rev_accts  num_rev_tl_bal_gt_0  num_sats  open_acc  pct_tl_nvr_dlq  \
0  9.0            4.0                  6.0       6.0       100.0            
1  32.0           9.0                  17.0      17.0      83.3             
2  18.0           4.0                  10.0      10.0      91.4             
3  7.0            5.0                  8.0       8.0       76.9             
4  8.0            4.0                  11.0      11.0      100.0            

   percent_bc_gt_75  pub_rec  pub_rec_bankruptcies             purpose  \
0  0.0               0.0      0.0                   debt_consolidation   
1  14.3              0.0      0.0                   credit_card          
2  100.0             0.0      0.0                   credit_card          
3  100.0             0.0      0.0                   car                  
4  100.0             0.0      0.0                   debt_consolidation   

   revol_bal  revol_util sub_grade  tax_liens        term  tot_coll_amt  \
0  138008.0   29.0        C1        0.0         60 months  0.0            
1  6133.0     31.6        A3        0.0         36 months  0.0            
2  16374.0    76.2        D1        0.0         60 months  0.0            
3  5753.0     100.9       D4        0.0         60 months  0.0            
4  16822.0    91.9        C3        0.0         36 months  0.0            

   

In [38]:
loans.shape

(1646801, 50)

We have a total of 50 features in the cleaned dataset.

### 05 - Writing data

Finally, we have to write this dataframe for further analysis:

In [30]:
loans.to_csv("/media/juanan/DATA/loan_data_analysis/data/clean/loans.csv.gz",
             compression='gzip', index=False)

We have reduced the disk size of the data from 1.4 GB to 132 MB.

In [39]:
!ls -lh /media/juanan/DATA/loan_data_analysis/data/clean/

total 132M
-rw-r--r-- 1 juanan juanan 132M jun 17 20:07 loans.csv.gz


In the next notebook, we are going to make an __Exploratory Data Analysis__ in order to get familiar with the data and understand the relationship between features.